OpenRA.Mods.Sp/Traits/World/CloudSpawner.cs

```diff
@@ -95,5 +104,73 @@ void SpawnCloud(Actor self)
 
 			self.World.AddFrameEndTask(w => w.Add(new Cloud(self.World, animation, startEdge, finishEdge, facing, info)));
 		}
+
+		void PreSpawnClouds(World world)
+		{
+			var facing = 256 * info.WindDirection / 32;
+			var direction = WAngle.FromFacing(facing);
+			var delta = new WVec(0, -1024, 0).Rotate(WRot.FromFacing(facing));
+
+			/*
+			 * The folloing codes find the middle point in map. Because clouds must
+			 * start from an edge to another edge, the paths of clouds
+			 * which contain this point will always be the longest path no matter the direction,
+			 * as long as the map is a symmetrical shape.
+			 *
+			 * For example, the middle point in map is "@", @ is (u = X/2, v = Y/2)
+			 * You can imagine any wind direction with all edges clouds possibly spawn.
+			 *
+			 *              ___X___
+             *              |	  |
+             *              |	  |
+             *              |  @  | Y
+             *              |	  |
+             *              |_____|
+			 *
+			 * By using this longest path, we can figure out the number of clouds should
+			 * spawn when a certain cloud completely goes over a longest path, which should be
+			 * neither too little nor too big compared with the clouds per map-cell later
+			 * spawned by `SpawnCloud`.
+			 */
+			var middlePoint = new MPos(world.Map.MapSize.X / 2, world.Map.MapSize.Y / 2);
+			var middlePointTarget = world.Map.CenterOfCell(middlePoint.ToCPos(world.Map));
+
+			// lDistance and averageSpeed used are for loop condition below.
+			var longestCloudDistance = System.Math.Abs(
+				(world.Map.DistanceToEdge(middlePointTarget, -delta) + info.Cordon).Length
+				+ (world.Map.DistanceToEdge(middlePointTarget, delta) + info.Cordon).Length);
+			var averageCloudSpeed = System.Math.Abs((int)info.Speed.Average(s => s.Length));
+
+			var stepPerSpawn = averageCloudSpeed * info.SpawnInterval;
+			var stepPerSpawnVector = stepPerSpawn * delta / 1024;
+
+			/*
+			 * Spawn clouds.
+			 *
+			 * Try to make clouds spawnning cover the entire map, meanwhile
+			 * with some randomlization. Choose random spawning point and
+			 * find startEdge, then add offset to let they go further to cover
+			 * the map. Offset will increase, just like the cloud already float

```

Comment:
The offset will be increased to simulate the movement the cloud would have made otherwise.